man/pgk.1
.\" generated with Ronn-NG/v0.9.1
.\" http://github.com/apjanke/ronn-ng/tree/0.9.1
.TH "PGK" "1" "July 2024" ""
.SH "NAME"
\fBpgk - Proxmox automator for deploy LXC and QEMU/KVM guests, with Cloud-init.\fR
.SH "SYNOPSIS"
pgk [\-b]\|\.\|\.\|\. [TYPE_GUEST] [IMAGE] [ID(S)] [IP(S)] [SHAPE]
.br
pgk [\-a]\|\.\|\.\|\. [TYPE_GUEST] [IMAGE] [SHAPE]
.br
pgk [\-t]\|\.\|\.\|\. [TYPE_GUEST] [IMAGE] [SHAPE]
.br
pgk [\-l]\|\.\|\.\|\. [TYPE_FILE]
.br
pgk [\-e]\|\.\|\.\|\. [TYPE_FILE] [NAMEFILE]
.br
pgk [\-e]\|\.\|\.\|\. [TYPE_FILE]
.br
pgk [\-u]\|\.\|\.\|\. [TYPE_FILE]
.SH "DESCRIPTION"
\fBPGK\fR aka \fBProxmox GK\fR, is a bash utility to automate deployment of LXC and QEMU/KVM guests, individually or in batches, via Cloud\-init\.
.SH "OPTIONS"
.SS "Batch Deployment"
\fIusage: pgk \-b [TYPE_GUEST] [IMAGE] [ID(S)] [IP(S)] [SHAPE]\fR
.P
\fB\-b\fR : Batch deployment of your LXC or QEMU/KVM guests from a same image\. pgk will define the appropriate parameters according to the image chosen, but it is still possible to pass a custom shape as an argument to adapt the deployment parameters\.
.IP "\[ci]" 4
\fB[TYPE_GUEST]\fR Required argument for select type of guests, string value lxc or kvm
.IP "\[ci]" 4
\fB[IMAGE]\fR Required argument for name of the image distribution, you can provide full namefile or use regex expression
.IP "\[ci]" 4
\fB[ID(S)]\fR Required argument for available CTID/VMID, numerical value like '100' or '100,101,103' for batch deployement
.IP "\[ci]" 4
\fB[IP(S)]\fR Required argument for static adress like '192\.168\.2\.100' or '192\.168\.2\.100,101,102' for batch deployement, default CIDR is provided form configuration file\.
.IP "\[ci]" 4
\fB[SHAPE]\fR Optional argument, specify a namefile for overide default guest settings
.IP "" 0
.SS "Automatic Deployment"
\fIusage: pgk \-a [TYPE_GUEST] [IMAGE] [SHAPE]\fR
.P
\fB\-a\fR : Minimalistic deployment option, pgk will define the appropriate parameters according to the image chosen, but it is still possible to pass a custom shape as an argument to adapt the deployment parameters\.
.IP "\[ci]" 4
\fB[TYPE_GUEST]\fR Required argument for select type of guests, string value lxc or kvm
.IP "\[ci]" 4
\fB[IMAGE]\fR Required argument for name of the image distribution, you can provide full namefile or use regex expression
.IP "\[ci]" 4
\fB[SHAPE]\fR Optional argument, specify a namefile for overide default guest settings
.IP "" 0
.SS "Template provining"
\fIusage: pgk \-t [TYPE_GUEST] [IMAGE] [SHAPE]\fR
.P
\fB\-t\fR : Provisionning a Proxmox template, the pre\-configuration of template required a shape file\.
.IP "\[ci]" 4
\fB[TYPE_GUEST]\fR Required argument for select type of guests, string value lxc or kvm
.IP "\[ci]" 4
\fB[IMAGE]\fR Required argument for name of the image distribution, you can provide full namefile or use regex expression
.IP "\[ci]" 4
\fB[SHAPE]\fR Required argument, specify a namefile for overide default guest settings
.IP "" 0
.SS "Standard Options"
\fIusage: pgk \-l [TYPE_FILE]\fR
.P
\fB\-l\fR : Option for List available LXC/QEMU images & configurations files
.IP "\[ci]" 4
\fB\-l img\fR list availability of guests images on pve host\.
.IP "\[ci]" 4
\fB\-l lxc\fR list shape preset available for LXC guests\.
.IP "\[ci]" 4
\fB\-l kvm\fR list shape preset available for QEMU/KVM guests\.
.IP "\[ci]" 4
\fB\-l ci\fR list generated cloud\-init configuration file\.
.IP "\[ci]" 4
\fB\-l fs\fR lists the expected required files and folders (debug)\.
.IP "" 0
.P
\fIusage: pgk \-u [TYPE_FILE]\fR
.P
\fB\-u\fR : Update configuration files\.
.IP "\[ci]" 4
\fBdshape\fR Regenerate all default shape
.IP "\[ci]" 4
\fBlxc_index\fR Manual update of the lxc index images
.IP "" 0
.P
\fIusage: pgk \-e [TYPE_FILE] [NAMEFILE]\fR
.br
\fIusage: pgk \-e [TYPE_FILE]\fR
.P
\fB\-e\fR : Option to edit or create a new shape (LXC/QEMU)\. But also a quick way to edit the main configuration file
.IP "\[ci]" 4
\fB\-e lxc [NAMEFILE]\fR Edit or create a shape preset file for lxc guest
.IP "\[ci]" 4
\fB\-e kvm [NAMEFILE]\fR Edit or create a shape preset file for QEMU/KVM guest
.IP "\[ci]" 4
\fB\-e config\fR Edit main configuration file\.
.IP "\[ci]" 4
\fB\-e kvm_index\fR Edit QEMU/KVM image index\.
.IP "\[ci]" 4
\fB\-e lxc_index\fR Edit LXC image index\.
.IP "" 0
.SH "EXAMPLES"
\fIEx: Batch deployment of three alpine Linux based virtual machines with a VMID (100,101,102) for each of them, and a dedicated IP address in the format 192\.168\.2\.***\. In this example, as no shape file is supplied, a default configuration will be applied and will perform a basic pre\-configuration of the guest: default user setup, OpenSSH configuration, qemu\-guest\-agent installation (only for KVM/QEMU guest), timezone and locales\.\fR
.P
\fBpgk \-b kvm alpine 100,101,102 192\.168\.2\.100,101,102\fR
.P
\fIEx: Minimalist deployment of Fedora 39 based virtual machine, the regex syntax is used to filter all images available\.\fR
.P
\fBpgk \-a kvm fedora\.*39\fR
.P
\fIEx: Creation of a "Proxmox template" based on alpine linux for an unprivileged container\. In this case, a shape preset is used to overwrite the default settings of container, and contains additionnaly directives for docker installation\.\fR
.P
\fBpgk \-t lxc alpine ss_ulxc_alpine\-docker\.yaml\fR
.P
\fIEx: Creation of a custom shape for a vm, I will then be asked to choose a pre\-filled base model depending on the distribution chosen\. The file is then edited using the default text editor to make any necessary changes\.\fR
.P
\fBpgk \-e kvm my_shape\.yaml\fR
.SH "FILES"
.IP "\[ci]" 4
\fB/etc/pgk/config\.cfg\fR
.br
the main configuration file
.IP "\[ci]" 4
\fB/var/lib/pgk\fR:
.br
Contains the basic/custom shapes for the different types of guest, and the indexes\. Also contains certain files needed to initialise cloud\-init for containers\.
.IP "\[ci]" 4
\fBinit/ds_kvm_*\.conf\fR: default shape for kvm guests
.IP "\[ci]" 4
\fBinit/ds_lxc_*\.conf\fR: default shape for lxc guests
.IP "\[ci]" 4
\fBinit/cinit\.sh\fR: payload required for cloud\-init configuration of containers\.
.IP "\[ci]" 4
\fBlxc_index\.json\fR: auto\-generated index, contains the addresses of the various container images and their default shape\.
.IP "\[ci]" 4
\fBkvm_index\.json\fR: index containing the addresses of the various QEMU/KVM guest images\.
.IP "" 0

.IP "\[ci]" 4
\fB/usr/lib/pgk\fR
.br
contains the scripts that execute the main functions\.
.IP "\[ci]" 4
\fB/usr/lib/pgk/pgk\.sh > /usr/bin/pgk\fR: Main executable, symlinked to /usr/bin/pgk\.
.IP "" 0

.IP "\[ci]" 4
\fB/var/log/pgk\.log\fR
.br
Supports tracking when guests are created, as well as the cloud\-init configuration process for containers\.
.IP "" 0
.SH "NOTES"
.SS "Configuration base storage"
The Proxmox storage model is very flexible, PGK needs to know where the different types of content are based on your Proxmox installation\.
.br
The /etc/pve/storage\.cfg file summarizes the status of all your storage\.
.br
Make sure you then adapt the variables for the ‘data store’ and ‘image store’ in the main configuration file /etc/pgk/config\.cfg\.
.P
\fIHS_PATH_DATASTR=/var/lib/vz # Full path of datastore\fR
.br
\fI# type of content required: iso,vztmpl,snippets\fR
.br
\fIHS_LOCALVM=local\-lvm # Name of guests image store (local\-zfs|local\-btrfs)\fR
.br
\fI# type of content required: rootdir,images\fR
.br
\fIHS_DATASTR=local # Common name of datastore\fR
.br
\fIHS_CIIMG=$HS_PATH_DATASTR/template/iso # Folder dedicated to QEMU/KVM cloud images\fR
.br
\fIHS_LXCIMG=$HS_PATH_DATASTR/template/cache # Folder dedicated to lxc images\fR
.br
\fIHS_PATH_CONF=$HS_PATH_DATASTR/snippets # Folder dedicated to cloud\-init config files (\.yaml)\fR
.SS "Set secret on environment variable"
PGK use your environment variable to store some secret
.IP "\[ci]" 4
\fBGS_CIUSER\fR: Default username, for all guests
.IP "\[ci]" 4
\fBGS_CIPASSWD_PLAIN\fR: Default plain password for GS_CIUSER, required only for OpenBSD (bcrypt exception)
.IP "\[ci]" 4
\fBGS_CIPASSWD_SHA\fR: Default hashed password for GS_CIUSER
.IP "" 0
.P
You must set all these variables before run this program :
.P
\fIecho "export GS_CIUSER='YOUR_USERNAME'" | tee \-a $HOME/\.bashrc\fR
.br
\fIecho "export GS_CIPASSWD_PLAIN='YOUR_PLAINTEXT_PASSWORD'" | tee \-a $HOME/\.bashrc\fR
.br
\fIecho "export GS_CIPASSWD_SHA='$(openssl passwd \-6 "YOUR_PLAINTEXT_PASSWORD")'" | tee \-a $HOME/\.bashrc\fR
.br
\fIsource $HOME/\.bashrc\fR
.SS "Custom shape file"
When you create a custom shape, the pre\-filled fields are made according to your parameters in the main configuration file \fB/etc/pgk/config\.cfg\fR\. All the shape files are structured in the same way and follow YAML syntax\. Three blocks must be present:
.IP "\[ci]" 4
\fBgs_values:\fR contains the values needed to create the lxc/qemu guest, they will be extracted and interpreted as environment variables, and only concern the creation of the vm or container (essentially interacting with \fB/usr/bin/pct\fR and \fB/usr/bin/qm\fR)\.
.IP "\[ci]" 4
\fBextra_guest_config:\fR this is an optional block, containing directives (one per line) that we'd like to pass directly into the guest's PVE configuration file, /etc/pve/lxc/CTID\.conf or /etc/pve/qemu\-server/VMID\.conf\. Use with caution, as the wrong instruction could prevent the guest from starting up correctly\.
.IP "\[ci]" 4
\fBcloud_init:\fR contains the directives that will be interpreted by Cloud\-init to configure the guest\.
.IP "" 0
.SH "BUGS"
You are welcome to submit bug reports via the PGK bug tracker (https://github\.com/asdeed/proxmox_gk/issues)\.
.SH "AUTHOR"
PGK\.1 written by Alexandre JAN \fIalexandre_jan@nodeswarm\.eu\fR
.SH "COPYRIGHT"
Copyright (C) \fI2023\-2024\fR Alexandre JAN
.P
This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or any later version\.


src/pgk.sh
#!/usr/bin/env bash
#
# Author : Alexandre JAN
# website : https://asded.fr
# Created : 18/12/2023
# Version : 0.1
# License : GPL-3.0 (GNU General Public License v3.0)
#
# This file is a main part of PGK.
#
# PGK is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# PGK is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with PGK.  If not, see <http://www.gnu.org/licenses/>.

##########################################################################################################
# DEBUG/TRACE                                                                     
##########################################################################################################

set -a
set -e
set -o errexit
set -o errtrace
set -o pipefail

source /etc/pgk/config.cfg
source "$HS_MAIN_SRCS/basic_mod.sh"
source "$HS_MAIN_SRCS/list_mod.sh"
source "$HS_MAIN_SRCS/edit_mod.sh"
source "$HS_MAIN_SRCS/set_mod.sh"
source "$HS_MAIN_SRCS/id_mod.sh"
source "$HS_MAIN_SRCS/make_mod.sh"
source "$HS_MAIN_SRCS/chk_mod.sh"

##########################################################################################################
# HELPER FUNCTIONS                                                                          
##########################################################################################################

helper() {
  echo ""
  msg_info "   ____                                       ____ _  __ "
  msg_info "  |  _ \ _ __ _____  ___ __ ___   _____  __  / ___| |/ / "
  msg_info "  | |_) | '__/ _ \ \/ / '_ ' _ \ / _ \ \/ / | |  _| ' /  "
  msg_info "  |  __/| | | (_) >  <| | | | | | (_) >  <  | |_| | . \  "
  msg_info "  |_|   |_|  \___/_/\_\_| |_| |_|\___/_/\_\  \____|_|\_\ "
  msg_info ""
  echo ""
  msg_syntax "Description:"
  echo "Proxmox automator for deploy LXC and QEMU/KVM guests, with Cloud-init." 
  echo ""
  msg_syntax "Usage:"
  echo "pgk [-b]... [TYPE_GUEST] [IMAGE] [ID(S)] [IP(S)] [SHAPE]"
  echo "pgk [-a]... [TYPE_GUEST] [IMAGE] [SHAPE]"
  echo "pgk [-t]... [TYPE_GUEST] [IMAGE] [SHAPE]"
  echo "pgk [-l]... [TYPE_FILE]"
  echo "pgk [-e]... [TYPE_FILE] [NAMEFILE]"
  echo "pgk [-e]... [TYPE_FILE]"
  echo "pgk [-u]... [TYPE_FILE]"
  echo ""
  msg_syntax "Options:"
  echo "-b batching deployement"
  echo "-a automated deploiement"
  echo "-t template provisioning"
  echo "-l listing configuration files and images"
  echo "-e editing configuration files"
  echo "-u updating configuration"
  echo ""
  msg_syntax "Arguments:"
  echo "[TYPE_GUEST]: Required argument for type of guests lxc|kvm"
  echo "[IMAGE]:      Name of the image distribution, accepts regex expressions"
  echo "[ID]:         Required argument for available CTID/VMID, numerical value like '100' or '100,101,103' for batch deployement"
  echo "[IP]:         Required argument for static adress like '192.168.2.100' or '192.168.2.100,101,102' for batch deployement"
  echo "[SHAPE]:      Optional argument, specify a shape namefile for overide default guest settings"
  echo "[TYPE_FILE]   Configuration files and/or fileset img|lxc|kvm|ci|fs|dshape|lxc_index|kvm_index"
  echo "[NAMEFILE]    Name of YAML shape file"
  echo ""
  msg_syntax "Example usage:"
  echo "Deployment of Fedora 39 based virtual machine, using regex syntax for filter all images available."
  msg_bold "    pgk -a kvm fedora.*39"
  echo "Batch deployment of three virtual machines based on alpine linux, with dedicated IP address"
  msg_bold "    pgk -b kvm alpine 100,101,102 192.168.2.100,101,102"
  echo "Make a "Proxmox template" based on alpine linux with custom configuration shape, for docker installation."
  msg_bold "    pgk -t lxc alpine ss_ulxc_alpine-docker.yaml"
  echo ""
  msg_info "Check manual for more information 'man pgk'"
  echo ""
}

##########################################################################################################
# NERVE OPTIONS                                                            
##########################################################################################################

# Check root 
if [[ $EUID -ne 0 ]]; then
  msg_error "This script must be run as root"
  exit 1
fi

# control boot 
log_command "CHECK-BOOT" "chk_lxc_index"

# control exec
pre_exec(){
  chk_pkg
  chk_local
  chk_local_lvm
  chk_ctrl_folders
  chk_ctrl_files
  chk_var_secret
  log_command "CHECK-EXEC" "clean_dshape" 
  log_command "CHECK-EXEC" "chk_guest_confiles" 
  log_command "CHECK-EXEC" "chk_lxc_shape"
  log_command "CHECK-EXEC" "chk_kvm_shape" 
}

# helper default display
if [[ $# == 0 ]]; then
    helper
else

# Main options for build guests
while getopts "b:a:t:l:e:u:" opt; do
    case $opt in
    b)
      pre_exec
      if [[ "$OPTARG" == "kvm" ]]; then
          count_args "$OPTARG" "$3" "$4" "$5" 
          chk_args distrib "$OPTARG" "$3"
          chk_args vmid "$OPTARG" "$4" 
          chk_args iplist "$OPTARG" "$5"
          GS_VMID="$4"
          GS_NETCONF="$5"
          check_vmid_list
          qm_getdistrib "$3"
          getshape kvm "$6" # optional used for overriding default value
          GS_SHAPECONF="$yaml_file"
          guest_qm_init
          msg_ok "KVM guest(s) initialised !"

      elif [[ "$OPTARG" == "lxc" ]]; then
          count_args "$OPTARG" "$3" "$4" "$5"          
          chk_args distrib "$OPTARG" "$3"
          chk_args vmid "$OPTARG" "$4" 
          chk_args iplist "$OPTARG" "$5"
          GS_VMID="$4"
          GS_NETCONF="$5"
          check_vmid_list          
          lxc_getdistrib "$3"
          getshape lxc "$6" # optional used for overdide defautl var.
          GS_SHAPECONF="$yaml_file"
          guest_lxc_init
          msg_ok "LXC guest(s) initialised !"
      else
         msg_error "'$OPTARG' is an invalid argument, you need to select a type for your guest: lxc|kvm"
      fi
      ;;
    a)
      pre_exec
      if [[ "$OPTARG" == "kvm" ]]; then
          chk_args distrib "$OPTARG" "$3"
          check_shape_yaml "$OPTARG" "$4"
          next_available_vmid
          GS_NETCONF="dhcp"
          qm_getdistrib "$3"
          getshape kvm "$4" # optional used for overriding default value
          GS_SHAPECONF="$yaml_file"
          guest_qm_init
          msg_ok "KVM guest(s) initialised !"

      elif [[ "$OPTARG" == "lxc" ]]; then
          chk_args distrib "$OPTARG" "$3"
          check_shape_yaml "$OPTARG" "$4"
          next_available_vmid
          GS_NETCONF="dhcp"
          lxc_getdistrib "$3"
          getshape lxc "$4" # optional used for overdide defautl var.
          GS_SHAPECONF="$yaml_file"
          guest_lxc_init
          msg_ok "LXC guest(s) initialised !"
      else
         msg_error "'$OPTARG' is an invalid argument, select type for your guest: lxc|kvm"
      fi
      ;;
    t)
      pre_exec
      if [[ "$OPTARG" == "kvm" ]]; then
          chk_args distrib "$OPTARG" "$3"
          check_shape_yaml "$OPTARG" "$4"
          next_available_vmid
          GS_NETCONF="dhcp"
          qm_getdistrib "$3"
          getshape kvm "$4" # needed for overriding default value
          GS_SHAPECONF="$yaml_file"
          QS_SBOOT="0" # Exception for covertion to template
          guest_qm_init
          guest_convert_tmpl qm
          msg_ok "KVM template created !"

      elif [[ "$OPTARG" == "lxc" ]]; then
          chk_args distrib "$OPTARG" "$3"
          check_shape_yaml "$OPTARG" "$4"
          next_available_vmid
          GS_NETCONF="dhcp"
          lxc_getdistrib "$3"
          getshape lxc "$4" # needed for overriding default value
          GS_SHAPECONF="$yaml_file"
          CS_SBOOT="0" # Exception for covertion to template
          guest_lxc_init
          guest_convert_tmpl lxc
          msg_ok "LXC template created !"
      else
         msg_error "'$OPTARG' is an invalid argument, select type for your guest: lxc|kvm"
      fi
      ;;
    l)
      if [[ "$OPTARG" == "img" ]]; then
        ls_qemu
        ls_lxc
      elif [[ "$OPTARG" == "kvm" ]]; then
        ls_qmshape
      elif [[ "$OPTARG" == "lxc" ]]; then
        ls_lxcshape
      elif [[ "$OPTARG" == "ci" ]]; then
        ls_ciconf
      elif [[ "$OPTARG" == "fs" ]]; then
        ls_fileset
        ls_folderset
      else
        msg_error "'$OPTARG' is an invalid argument, use one of the following: img|kvm|lxc|ci|fs"
      fi
      ;;
    e)
      if [[ "$OPTARG" == "kvm" ]]; then
        if [ -z "$3" ]; then
          msg_error "Error: your kvm shape require namefile" >&2
          exit 1
        else 
          check_shape_yaml "$OPTARG" "$3"
          edit_qemu_shape "$3"
        fi
      elif [[ "$OPTARG" == "lxc" ]]; then
        if [ -z "$3" ]; then
          msg_error "Error: your lxc shape require namefile" >&2
          exit 1
        else 
          check_shape_yaml "$OPTARG" "$3"
          edit_lxc_shape "$3"
        fi
      elif [[ "$OPTARG" == "config" ]]; then
        edit_confile
      elif [[ "$OPTARG" == "kvm_index" ]]; then
        edit_index kvm
      elif [[ "$OPTARG" == "lxc_index" ]]; then
        edit_index lxc
      else
         msg_error "'$OPTARG' is an invalid argument, use one of the following: kvm|lxc|config|kvm_index|lxc_index"
      fi
      ;;
    u)
      if [[ "$OPTARG" == "dshape" ]]; then
        log_command "CHECK-CLEAN_DSHAPE" "clean_dshape"
      elif [[ "$OPTARG" == "lxc_index" ]]; then
        chk_index_lxc_update
      # TODO function for update cloud-init config guests
      else
        msg_error "'$OPTARG' is an invalid argument, use one of the following: dshape|lxc_index"
      fi
      ;;
    *)
      msg_error "Argument needed for $1 option"
        if [[ "$1" == "-l" ]]; then
          msg_info "Use one of the following arguments: img|kvm|lxc|ci|fs"
        elif [[ "$1" == "-b" ]]; then
          msg_info "Use one of the following arguments: lxc|kvm"
        elif [[ "$1" == "-a" ]]; then
          msg_info "Use one of the following arguments: lxc|kvm"
        elif [[ "$1" == "-t" ]]; then
          msg_info "Use one of the following arguments: lxc|kvm"
        elif [[ "$1" == "-e" ]]; then
          msg_info "Use one of the following arguments: kvm|lxc|config|kvm_index|lxc_index"
        elif [[ "$1" == "-u" ]]; then
          msg_info "Use the following arguments: dshape|lxc_index"
        fi
      exit 1
      ;;
    esac
done
fi
